src/Hadrian/Haskell/Cabal/Type.hs

```diff
@@ -1,23 +1,24 @@
 module Hadrian.Haskell.Cabal.Type where
 
 import Development.Shake.Classes
+import Distribution.PackageDescription (GenericPackageDescription, PackageDescription)
+import GHC.Generics
 import Hadrian.Package.Type
 
--- TODO: Use fine-grain tracking instead of tracking the whole Cabal file.
 -- | Haskell package metadata extracted from a Cabal file.
 data Cabal = Cabal
-    { dependencies :: [PackageName]
-    , name         :: PackageName
-    , synopsis     :: String
-    , version      :: String
-    } deriving (Eq, Read, Show, Typeable)
+    { name                      :: PackageName
+    , version                   :: String
+    , synopsis                  :: String
+    , genericPackageDescription :: GenericPackageDescription
+    , packageDescription        :: PackageDescription
+    , packageDependencies       :: [Package]
+    } deriving (Eq, Show, Typeable, Generic)
 
-instance Binary Cabal where
-    put = put . show
-    get = fmap read get
+instance Binary Cabal
 
 instance Hashable Cabal where
     hashWithSalt salt = hashWithSalt salt . show
 
 instance NFData Cabal where
-    rnf (Cabal a b c d) = a `seq` b `seq` c `seq` d `seq` ()
+    rnf (Cabal a b c d e f) = a `seq` b `seq` c `seq` d `seq` e `seq` f `seq` ()
```

Comment:
use derivation here as well?

Reply:
Turns out we can derive `NFData` but not `Hashable` because `GenericPackageDescription` does not have a `Hashable` instance. Will be done in an upcoming commit.